driver-core/src/main/com/mongodb/internal/connection/ServerAddressWithResolver.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.MongoSocketException;
+import com.mongodb.ServerAddress;
+import com.mongodb.lang.Nullable;
+import com.mongodb.spi.dns.InetAddressResolver;
+import com.mongodb.spi.dns.InetAddressResolverProvider;
+
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+final class ServerAddressWithResolver extends ServerAddress {
+    private static final long serialVersionUID = -1;  // this internal class will never be serialized
+
+    private static final InetAddressResolver DEFAULT_INET_ADDRESS_PROVIDER;
+
+    static {
+        DEFAULT_INET_ADDRESS_PROVIDER = StreamSupport.stream(ServiceLoader.load(InetAddressResolverProvider.class).spliterator(), false)
+                .findFirst()
+                .map(InetAddressResolverProvider::create)
+                .orElse(null);
+    }
+
+    private final transient InetAddressResolver resolver;
+
+    ServerAddressWithResolver(final ServerAddress serverAddress, @Nullable final InetAddressResolver inetAddressResolver) {
+        super(serverAddress.getHost(), serverAddress.getPort());
+        this.resolver = inetAddressResolver == null ? DEFAULT_INET_ADDRESS_PROVIDER : inetAddressResolver;
+    }
+
+    @Override
+    public InetSocketAddress getSocketAddress() {
+        if (resolver == null) {
+            return super.getSocketAddress();
+        }
+
+        return getSocketAddresses().get(0);
+    }
+
+    @Override
+    public List<InetSocketAddress> getSocketAddresses() {
+        if (resolver == null) {
+            return super.getSocketAddresses();
+        }
+        try {
+            return resolver.lookupByName(getHost())
+                    .stream()
+                    .map(inetAddress -> new InetSocketAddress(inetAddress, getPort())).collect(Collectors.toList());
+        } catch (UnknownHostException e) {
+            throw new MongoSocketException(e.getMessage(), this, e);
+        }
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        return super.equals(o);
+    }
+
+    @Override
+    public int hashCode() {
+        return super.hashCode();
+    }
```

Comment:
1. Implementing overrides like this is equivalent to not overriding at all. The latter approach is taken for the `toString` method.
2. With the way `ServerAddress.equals` is implemented, no `ServerAddressWithResolver` is ever equal to an object whose runtime type is `ServerAddress` (and wise versa, because the current `equals` implementation is symmetric, as it must be). Is this actually OK, or do we have sets / maps / something else where we rely on equality and objects of these two types may be compared with each other?

Reply:
Done

Reply:
We have `ConcurrentMap<ServerAddress, ServerTuple> AbstractMultiServerCluster.addressToServerTupleMap` and use `Set<ServerAddress>` as the type of a local variable / return value in multiple places. Are we sure that instances of `ServerAddress` and instances of `ServerAddressWithResolver` are never mixed together? Even if don't mix them as the contents of the aforementioned collections, it's enough to read from them (`get` from map, or `contains` on a set) using an instance of the type different from the one stored inside, to get ourselves into trouble.

Reply:
Is your suggestion to remove equal/hashCode here?  If so, I think it's fine.  Either way, I don't anticipate that these methods will ever be called by the driver or Stream implementation of an instance of this class.

Reply:
Note that this class only exists because the StreamFactory interface is public.  It it wasn't we would have much more flexibility in the design.

Reply:
> Is your suggestion to remove equal/hashCode here?

I did not have a suggestion, I was just interested in whether we are sure that instances of `ServerAddress` and `ServerAddressWithResolver` are never used together in a context where `hashCode`/`equals` are used.